proposals/interface-types/working-notes/Types.md

```diff
@@ -37,36 +62,158 @@ Integer values.
+### Structured Value Types
 
-## `array`
+Structured types allow the representation of more complex combinations of
+values. There are five forms of structured type: arrays, sequences, records,
+protocols and algebraic variants.
 
-The `array` interface value type is a compound type defined by the following abstract syntax:
+```
+structuretype ::= optiontype | eithertype |
+   arraytype | sequencetype | recordtype | varianttype
+```
+
+#### Option Type
+
+Option types are used to model nullability in types.
 
 ```
-#array ::= array #interface-valtype
+optiontype ::= option type
 ```
 
-An `array` value is a homogenous sequence of interface values with type `$interface-value-type`.
+The `option` type can be viewed as a use of type variants -- with one variant
+being the enumerated symbol `$none` and the other being the single value wrapped
+in a `$some` variant.
+
+#### Either Type
+
+The `either` type is used to model situations where one of two values may be
+returned by a function. A classic case for this is in modeling error returns in
+APIs.
+
+```
+eithertype ::= either type
+```
+
+#### Arrays
+
+An `array` consists of a finite vector of elements, each of which has the same
+type.
+
+```
+arraytype ::= array type
+```
+
+Both the type and the representation of each element of the array is assumed to
+be identical. Furthermore, arrays are assumed to have a know element
+count. (Indexing into arrays is not part of this specification.)
+
+#### Sequences
+
+A sequence consists of an ordered collection of elements; each of which has the
+same type.
+
+Sequence elements may not have identical representation; nor is it necessarily
+known how many elements there are in a sequence.
+
+```
+sequencetype ::= sequence type
+```
+
+#### Records
+
+A `record` type is a tuple composition of types. 
+
+```
+#record ::= record [ vec(interfacetype) ]
+```
+
+Fields in a record are accessed via their index; for convenience, field indices
+may be given a local name:
+
+```
+record [ $name string $age u64]
+```
+
+#### Variant Types
+
+A `variant` consists of an ordered collection of alternative vectors of types.
+
+```
+varianttype ::= oneof [ vec(vec(interfacetype)) ]
+```
+
+Variants are accessed by their index in the vector of alternatives. For
+convenience, variants may be given a local name:
+
+```
+variant [ $id [s64] $name [string] ]
+```
+
+A given variant may not have any type arguments; in which case it corresponds to
+an enumerated symbol. For example, the boolean type may be viewed as a synonym
+for:
+
+```
+(type $boolean (variant ($false) ($true)))
+```
+
+Additionally, the option type may be viewed as a synonym for:
+
+```
+(type ($option #tp) (variant ($nil) ($some #tp)))
+```
+
+>Note: this definition is not technically legal.
+
+The representation of a variant is determined by the representation of each of
+its alternatives.
+
+### Function Type
+
+A `functionType` denotes a function whose signature is expressed as IT types.
```

Comment:
nit: I think `IT types` could be reworded as `interface types` and still have the same meaning.